src/Workspaces/Core/Portable/Shared/Utilities/ProducerConsumer.cs

```diff
@@ -79,16 +79,64 @@ internal static class ProducerConsumer<TItem>
     /// </summary>
     public static Task RunAsync<TArgs>(
         ProducerConsumerOptions options,
-        Func<Action<TItem>, TArgs, Task> produceItems,
-        Func<IAsyncEnumerable<TItem>, TArgs, Task> consumeItems,
+        Func<Action<TItem>, TArgs, CancellationToken, Task> produceItems,
+        Func<IAsyncEnumerable<TItem>, TArgs, CancellationToken, Task> consumeItems,
         TArgs args,
         CancellationToken cancellationToken)
     {
         return RunImplAsync(
             options,
-            static (onItemFound, args) => args.produceItems(onItemFound, args.args),
-            static (reader, args) => args.consumeItems(reader.ReadAllAsync(args.cancellationToken), args.args),
-            (produceItems, consumeItems, args, cancellationToken),
+            static (onItemFound, args, cancellationToken) => args.produceItems(onItemFound, args.args, cancellationToken),
+            static (reader, args, cancellationToken) => args.consumeItems(reader.ReadAllAsync(cancellationToken), args.args, cancellationToken),
+            (produceItems, consumeItems, args),
+            cancellationToken);
+    }
+
+    /// <summary>
+    /// Version of RunAsync that will process <paramref name="source"/> in parallel.
+    /// </summary>
+    public static Task RunParallelAsync<TSource, TArgs>(
+        IEnumerable<TSource> source,
+        Func<TSource, Action<TItem>, TArgs, CancellationToken, Task> produceItems,
+        Func<IAsyncEnumerable<TItem>, TArgs, CancellationToken, Task> consumeItems,
+        TArgs args,
+        CancellationToken cancellationToken)
+    {
+        return RunAsync(
+            // We're running in parallel, so we def have multiple writers
+            ProducerConsumerOptions.SingleReaderOptions,
+            produceItems: static (callback, args, cancellationToken) =>
```

Comment:
> produceItems: static (callback, args, cancellationToken) => [](http://example.com/codeflow?start=11&length=60)

I think this might be one level too deep for me to understand without having had more coffee. Local function for the producer here might make things clearer.

Reply:
Have coffee!  I can also extract into local functions on a follow-up if that's ok.